File: backend/LimsAuth.Api/Models/Schedule/Reservation.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace LimsAuth.Api.Models;

[Table("reservations")]
public class Reservation
{
    [Key]
    [Column("id")]
    public Guid Id { get; set; } = Guid.NewGuid();

    [Required]
    [Column("semester_id")]
    public Guid SemesterId { get; set; }

    [Required]
    [Column("lab_id")]
    public Guid LabId { get; set; }

    [Column("use_date")]
    public DateTime UseDate { get; set; }

    [Column("day_of_week")]
    public int DayOfWeek { get; set; }

    [Column("period_numbers")]
    [MaxLength(200)]
    public string? PeriodNumbersJson { get; set; }

    [NotMapped]
    public List<int> PeriodNumbers
    {
        get => string.IsNullOrEmpty(PeriodNumbersJson)
            ? new List<int>()
            : System.Text.Json.JsonSerializer.Deserialize<List<int>>(PeriodNumbersJson) ?? new List<int>();
        set => PeriodNumbersJson = System.Text.Json.JsonSerializer.Serialize(value);
    }

    [Column("week_number")]
    public int WeekNumber { get; set; }

    [Column("expected_duration_hours")]
    public double? ExpectedDurationHours { get; set; }

    [Required]
    [Column("project_name")]
    [MaxLength(200)]
    public string ProjectName { get; set; } = string.Empty;

    [Column("project_category")]
    [MaxLength(50)]
    public string ProjectCategory { get; set; } = string.Empty;

    [Column("remark")]
    [MaxLength(1000)]
    public string? Remark { get; set; }

    [Required]
    [Column("applicant_id")]
    public Guid ApplicantId { get; set; }

    [Required]
    [Column("applicant_name")]
    [MaxLength(100)]
    public string ApplicantName { get; set; } = string.Empty;

    [Required]
    [Column("applicant_phone")]
    [MaxLength(20)]
    public string ApplicantPhone { get; set; } = string.Empty;

    [Column("project_leader_id")]
    public Guid? ProjectLeaderId { get; set; }

    [Column("project_leader_name")]
    [MaxLength(100)]
    public string? ProjectLeaderName { get; set; }

    [Column("project_leader_phone")]
    [MaxLength(20)]
    public string? ProjectLeaderPhone { get; set; }

    [Column("member_grade")]
    [MaxLength(50)]
    public string? MemberGrade { get; set; }

    [Column("member_class_id")]
    public Guid? MemberClassId { get; set; }

    [Column("member_class_name")]
    [MaxLength(100)]
    public string? MemberClassName { get; set; }

    [Column("member_count")]
    public int? MemberCount { get; set; }

    [Column("status")]
    [MaxLength(20)]
    public ApprovalStatus Status { get; set; } = ApprovalStatus.Pending;

    [Column("approval_comment")]
    [MaxLength(500)]
    public string? ApprovalComment { get; set; }

    [Column("approved_by")]
    public Guid? ApprovedBy { get; set; }

    [Column("approved_at")]
    public DateTime? ApprovedAt { get; set; }

    [Column("is_cancelled")]
    public bool IsCancelled { get; set; }

    [Column("cancel_reason")]
    [MaxLength(500)]
    public string? CancelReason { get; set; }

    [Column("cancelled_by")]
    public Guid? CancelledBy { get; set; }

    [Column("cancelled_at")]
    public DateTime? CancelledAt { get; set; }

    [Column("created_at")]
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    [Column("created_by")]
    public string? CreatedBy { get; set; }

    [Column("updated_at")]
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    [Column("updated_by")]
    public string? UpdatedBy { get; set; }

    public virtual Semester? Semester { get; set; }
    public virtual Lab? Lab { get; set; }
    public virtual User? Applicant { get; set; }
}

File: mobile/lib/models/reservation.dart
import 'lab.dart';
import '../utils/num_utils.dart';
import 'user.dart';

class Reservation {
  final String id;
  final String? semesterId;
  final String? semesterName;
  final String? labId;
  final String? labName;
  final String? useDate;
  final int dayOfWeek;
  final List<int>? periodNumbers;
  final int weekNumber;
  final double expectedDurationHours;
  final String? projectName;
  final String? projectCategory;
  final String? remark;
  final String? applicantId;
  final String? applicantName;
  final String? applicantPhone;
  final String? projectLeaderId;
  final String? projectLeaderName;
  final String? projectLeaderPhone;
  final String? memberGrade;
  final String? memberClassId;
  final String? memberClassName;
  final int memberCount;
  final String? status;
  final String? approvalComment;
  final String? approvedBy;
  final String? approverName;
  final String? approvedAt;
  final bool isCancelled;
  final String? cancelReason;
  final String? createdAt;
  final Lab? lab;

  Reservation({
    required this.id,
    this.semesterId,
    this.semesterName,
    this.labId,
    this.labName,
    this.useDate,
    this.dayOfWeek = 0,
    this.periodNumbers,
    this.weekNumber = 0,
    this.expectedDurationHours = 0,
    this.projectName,
    this.projectCategory,
    this.remark,
    this.applicantId,
    this.applicantName,
    this.applicantPhone,
    this.projectLeaderId,
    this.projectLeaderName,
    this.projectLeaderPhone,
    this.memberGrade,
    this.memberClassId,
    this.memberClassName,
    this.memberCount = 0,
    this.status,
    this.approvalComment,
    this.approvedBy,
    this.approverName,
    this.approvedAt,
    this.isCancelled = false,
    this.cancelReason,
    this.createdAt,
    this.lab,
  });

  factory Reservation.fromJson(Map<String, dynamic> json) {
    return Reservation(
      id: json['id']?.toString() ?? '',
      semesterId: json['semesterId']?.toString(),
      semesterName: json['semesterName']?.toString(),
      labId: json['labId']?.toString(),
      labName: json['labName']?.toString(),
      useDate: json['useDate']?.toString(),
      dayOfWeek: safeInt(json['dayOfWeek']),
      periodNumbers: (json['periodNumbers'] as List<dynamic>?)?.cast<int>(),
      weekNumber: safeInt(json['weekNumber']),
      expectedDurationHours: safeDouble(json['expectedDurationHours']),
      projectName: json['projectName']?.toString(),
      projectCategory: json['projectCategory']?.toString(),
      remark: json['remark']?.toString(),
      applicantId: json['applicantId']?.toString(),
      applicantName: json['applicantName']?.toString(),
      applicantPhone: json['applicantPhone']?.toString(),
      projectLeaderId: json['projectLeaderId']?.toString(),
      projectLeaderName: json['projectLeaderName']?.toString(),
      projectLeaderPhone: json['projectLeaderPhone']?.toString(),
      memberGrade: json['memberGrade']?.toString(),
      memberClassId: json['memberClassId']?.toString(),
      memberClassName: json['memberClassName']?.toString(),
      memberCount: safeInt(json['memberCount']),
      status: json['status']?.toString(),
      approvalComment: json['approvalComment']?.toString(),
      approvedBy: json['approvedBy']?.toString(),
      approverName: json['approverName']?.toString(),
      approvedAt: json['approvedAt']?.toString(),
      isCancelled: json['isCancelled'] as bool? ?? false,
      cancelReason: json['cancelReason']?.toString(),
      createdAt: json['createdAt']?.toString(),
      lab: json['lab'] != null
          ? Lab.fromJson(json['lab'] as Map<String, dynamic>)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'semesterId': semesterId,
        'labId': labId,
        'useDate': useDate,
        'dayOfWeek': dayOfWeek,
        'periodNumbers': periodNumbers,
        'weekNumber': weekNumber,
        'expectedDurationHours': expectedDurationHours,
        'projectName': projectName,
        'projectCategory': projectCategory,
        'remark': remark,
        'projectLeaderId': projectLeaderId,
        'projectLeaderName': projectLeaderName,
        'projectLeaderPhone': projectLeaderPhone,
        'memberGrade': memberGrade,
        'memberClassId': memberClassId,
        'memberClassName': memberClassName,
        'memberCount': memberCount,
      };
}

class CreateReservationRequest {
  final String semesterId;
  final String labId;
  final String? useDate;
  final int dayOfWeek;
  final List<int>? periodNumbers;
  final int weekNumber;
  final double expectedDurationHours;
  final String projectName;
  final String? projectCategory;
  final String? remark;
  final String? projectLeaderId;
  final String? projectLeaderName;
  final String? projectLeaderPhone;
  final String? memberGrade;
  final String? memberClassId;
  final String? memberClassName;
  final int memberCount;

  CreateReservationRequest({
    required this.semesterId,
    required this.labId,
    this.useDate,
    this.dayOfWeek = 0,
    this.periodNumbers,
    this.weekNumber = 0,
    this.expectedDurationHours = 0,
    required this.projectName,
    this.projectCategory,
    this.remark,
    this.projectLeaderId,
    this.projectLeaderName,
    this.projectLeaderPhone,
    this.memberGrade,
    this.memberClassId,
    this.memberClassName,
    this.memberCount = 0,
  });

  Map<String, dynamic> toJson() => {
        'semesterId': semesterId,
        'labId': labId,
        'useDate': useDate,
        'dayOfWeek': dayOfWeek,
        'periodNumbers': periodNumbers,
        'weekNumber': weekNumber,
        'expectedDurationHours': expectedDurationHours,
        'projectName': projectName,
        'projectCategory': projectCategory,
        'remark': remark,
        'projectLeaderId': projectLeaderId,
        'projectLeaderName': projectLeaderName,
        'projectLeaderPhone': projectLeaderPhone,
        'memberGrade': memberGrade,
        'memberClassId': memberClassId,
        'memberClassName': memberClassName,
        'memberCount': memberCount,
      };
}

class ApprovalRequest {
  final String? comment;
  final String? approverName;

  ApprovalRequest({this.comment, this.approverName});

  Map<String, dynamic> toJson() => {
        'comment': comment,
        'approverName': approverName,
      };
}
